File: mindspore/python/mindspore/_extends/graph_kernel/expanders/softmax_cross_entropy_with_logits.py
# Copyright 2021 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ===========================================================================
"""generate json desc for SoftmaxCrossEntropyWithLogits"""
from mindspore._extends.graph_kernel.model.model import DataFormat as DF
from ._utils import Expander, ExpanderInfoValidator as VLD


@VLD.add_format(DF.DEFAULT, DF.DEFAULT)
class SoftmaxCrossEntropyWithLogits(Expander):
    """SoftmaxCrossEntropyWithLogits expander"""

    def _expand(self, graph_builder):
        logits, label = self.inputs
        # Calculate softmax_cross_entropy_with_logits(logits, label)
        # formula of softmax_cross_entropy_with_logits is : -reduce_sum(label * log(softmax(logits)))
        axis = (-1,)
        max_x = graph_builder.emit('ReduceMax', [logits], attrs={'reduce_axis': axis, 'keep_dims': True})
        data_sub = graph_builder.emit('Sub', [logits, max_x])
        data_exp = graph_builder.emit('Exp', [data_sub])
        data_expsum = graph_builder.emit('ReduceSum', [data_exp], attrs={'reduce_axis': axis, 'keep_dims': True})
        data_softmax = graph_builder.emit('RealDiv', [data_exp, data_expsum])
        const_eps = graph_builder.value(logits.dtype, 0.000001)
        data_softmax_safety = graph_builder.emit("Maximum", [data_softmax, const_eps])
        softmax_log = graph_builder.emit('Log', [data_softmax_safety])
        label_mul_log = graph_builder.emit('Mul', [label, softmax_log])
        tmp_res = data_expsum = graph_builder.emit('ReduceSum', [label_mul_log], attrs={
            'reduce_axis': axis, 'keep_dims': False})
        loss = graph_builder.emit('Neg', [tmp_res])
        dlogits = graph_builder.emit('Sub', [data_softmax, label])
        return loss, dlogits

File: mindspore/ccsrc/plugin/device/cpu/kernel/nnacl/fp32_grad/softmax_cross_entropy_with_logits.c
/**
 * Copyright 2022 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "nnacl/fp32_grad/softmax_cross_entropy_with_logits.h"
#include <math.h>

void ForwardPostExecute(const float *labels, const float *logits, float *grads, float *output2,
                        size_t number_of_classes, int batch_size) {
  float eps = 1e-6;
  if (grads != NULL) {
    for (size_t i = 0; i < (size_t)(batch_size); ++i) {
      float loss = 0.f;
      for (size_t j = 0; j < number_of_classes; ++j) {
        float logit = -logf(logits[i * number_of_classes + j] <= 0.0 ? eps : logits[i * number_of_classes + j]);
        grads[i * number_of_classes + j] = (logits[i * number_of_classes + j] - labels[i * number_of_classes + j]);
        loss += labels[i * number_of_classes + j] * logit;
      }
      output2[i] = loss;
    }
  } else {
    for (size_t i = 0; i < (size_t)(batch_size); ++i) {
      float loss = 0.f;
      for (size_t j = 0; j < number_of_classes; ++j) {
        float logit = -logf(logits[i * number_of_classes + j] <= 0.0 ? eps : logits[i * number_of_classes + j]);
        loss += labels[i * number_of_classes + j] * logit;
      }
      output2[i] = loss;
    }
  }
}
